src/lib/concurrency/transaction_manager.cpp

```diff
@@ -23,7 +24,40 @@ TransactionManager::TransactionManager()
 CommitID TransactionManager::last_commit_id() const { return _last_commit_id; }
 
 std::shared_ptr<TransactionContext> TransactionManager::new_transaction_context() {
-  return std::make_shared<TransactionContext>(_next_transaction_id++, _last_commit_id);
+  const TransactionID snapshot_commit_id = _last_commit_id;
+  return std::make_shared<TransactionContext>(_next_transaction_id++, snapshot_commit_id);
+}
+
+void TransactionManager::register_transaction(const CommitID snapshot_commit_id) {
+  std::unique_lock<std::mutex> lock(_mutex_active_snapshot_commit_ids);
+  _active_snapshot_commit_ids.insert(snapshot_commit_id);
+}
+
+void TransactionManager::deregister_transaction(const CommitID snapshot_commit_id) {
+  std::unique_lock<std::mutex> lock(_mutex_active_snapshot_commit_ids);
+
+  auto it =
+      std::find(std::begin(_active_snapshot_commit_ids), std::end(_active_snapshot_commit_ids), snapshot_commit_id);
+
+  if (it != std::end(_active_snapshot_commit_ids)) {
+    _active_snapshot_commit_ids.erase(it);
+    return;
+  }
+
+  Fail(opossum::trim_source_file_path(__FILE__) + ":" BOOST_PP_STRINGIZE(__LINE__) +
+       " Could not find snapshot_commit_id in TransactionManager's _active_snapshot_commit_ids. Therefore," +
+       " the removal failed and the function should not have been called.");
+}
+
+CommitID TransactionManager::get_lowest_active_snapshot_commit_id() const {
+  std::unique_lock<std::mutex> lock(_mutex_active_snapshot_commit_ids);
+
+  if (_active_snapshot_commit_ids.empty()) {
+    return MvccData::MAX_COMMIT_ID;
```

Comment:
Maybe `std::nullopt` (and `std::optional<CommitID>` as the function return type), instead of a somewhat magic number?